### PR TITLE
Insert delta of current_state_events to be more efficient

### PR DESCRIPTION
**Pull request by erikjohnston**

### CODE REVIEW

File: synapse/storage/events.py [hunk 1]
```diff
@@ -476,37 +476,74 @@ def _persist_events_txn(self, txn, events_and_contexts, backfilled,
...
-                        "event_id": ev_id,
-                        "room_id": room_id,
-                        "type": key[0],
-                        "state_key": key[1],
-                    }
-                    for key, ev_id in current_state.iteritems()
-                ],
-            )
+            # Figure out what has changed (if anything). Then we simply delete
+            # and readd the keys that have been changed.
```
Comment (NegativeMjark) [inline synapse/storage/events.py]: "re-add" ?

File: synapse/storage/events.py [hunk 2]
```diff
@@ -476,37 +476,74 @@ def _persist_events_txn(self, txn, events_and_contexts, backfilled,
...
-                table="current_state_resets",
-                values={"event_stream_ordering": max_stream_order}
-            )
-
-            self._simple_delete_txn(
+            existing_state_rows = self._simple_select_list_txn(
                 txn,
                 table="current_state_events",
                 keyvalues={"room_id": room_id},
+                retcols=["event_id", "type", "state_key"],
```
Comment (NegativeMjark) [inline synapse/storage/events.py]: Why are you returning the "type" and "state_key" here?

Comment (erikjohnston) [inline synapse/storage/events.py]: Because we're using it below?

Comment (NegativeMjark) [inline synapse/storage/events.py]: Oh I failed to read that far down the file :(

File: synapse/storage/events.py [hunk 3]
```diff
@@ -406,6 +372,91 @@ def _calculate_new_extremeties(self, room_id, events):
...
 
     @defer.inlineCallbacks
+    def _calculate_state_delta(self, room_id, events_context, new_latest_event_ids):
+        """Calculate the new state deltas for a room.
+
+        Assumes that we are only persisting events for one room at a time.
+
+        Returns:
+            2-tuple (to_delete, to_insert) where both are state dicts, i.e.
+            (type, state_key) -> event_id. `to_delete` are the entreis to
```
Comment (NegativeMjark) [inline synapse/storage/events.py]: entries

**Pull request merged**